src/tools/run-make-support/src/lib.rs

```diff
@@ -0,0 +1,91 @@
+    /// We maintain a snapshot mapping of environment variables and their values.
+    pub(crate) vars: HashMap<String, String>,
+}
+
+impl RmakeSupportCtxt {
+    /// Initialize the `rmake-test-v2` support context with a snapshot of the current process'
+    /// environment.
+    pub fn init() -> Self {
+        let mut vars = std::env::vars().collect::<HashMap<_, _>>();
+        let ld_lib_path_envvar = vars.get("LD_LIB_PATH_ENVVAR").unwrap();
+        let mut target_rpath_env_path = String::new();
+        target_rpath_env_path.push_str(vars.get("TMPDIR").unwrap());
+        target_rpath_env_path.push(':');
+        target_rpath_env_path.push_str(vars.get(ld_lib_path_envvar).unwrap());
+        vars.insert("TARGET_RPATH_ENV".to_string(), target_rpath_env_path);
+        Self { vars }
+    }
+
+    pub fn tmpdir(&self) -> &str {
+        self.vars.get("TMPDIR").unwrap()
+    }
+
+    pub fn rustc(&self) -> &str {
+        self.vars.get("RUSTC").unwrap()
+    }
+
+    fn run_common(&self, bin_name: String) -> Result<Output, String> {
+        let mut bin_path = PathBuf::new();
+        bin_path.push(self.tmpdir());
+        bin_path.push(&bin_name);
+        let output = Command::new(bin_path)
+            .env("TARGET_RPATH_ENV", self.vars.get("TARGET_RPATH_ENV").unwrap())
+            .env(self.vars.get("LD_LIB_PATH_ENVVAR").unwrap(), {
+                let ld_lib_path_envvar = self.vars.get("LD_LIB_PATH_ENVVAR").unwrap();
+                let mut target_rpath_env_path = String::new();
+                target_rpath_env_path.push_str(self.tmpdir());
+                target_rpath_env_path.push(':');
+                target_rpath_env_path.push_str(self.vars.get("TARGET_RPATH_DIR").unwrap());
+                target_rpath_env_path.push(':');
+                target_rpath_env_path.push_str(self.vars.get(ld_lib_path_envvar).unwrap());
+                target_rpath_env_path
+            })
+            .output()
+            .map_err(|e| format!("{:?}", e))?;
+        Ok(output)
+    }
+
+    /// Run a built binary and make sure it succeeds.
+    pub fn run(&self, bin_name: String) -> Result<Output, String> {
+        let output = self.run_common(bin_name)?;
+        assert!(output.status.success());
+        Ok(output)
+    }
+
+    /// Run a built binary and make sure it fails.
+    pub fn run_fail(&self, bin_name: String) -> Result<Output, String> {
+        let output = self.run_common(bin_name)?;
+        assert!(!output.status.success());
+        Ok(output)
+    }
+}
+
+#[macro_export]
+macro_rules! rustc {
+    ($scx: expr, $args: expr) => {{
+        let sh = $crate::xshell::Shell::new().unwrap();
+        let rustc = $scx.rustc();
+        let args = $args;
+        let mut cmd = $crate::xshell::cmd!(sh, "{rustc}")
+            .arg("--out-dir")
+            .arg($scx.tmpdir())
+            .arg("-L")
+            .arg($scx.tmpdir());
+        for arg in args.split_whitespace() {
+            cmd = cmd.arg(arg);
+        }
```

Comment:
Maybe do:

```suggestion
        let mut cmd = $crate::xshell::cmd!(sh, concat_str!("{rustc} ", $args))
            .arg("--out-dir")
            .arg($scx.tmpdir())
            .arg("-L")
            .arg($scx.tmpdir());
```

This should allow interpolations the same way as xshell itself and should correctly handle quoting arguments. I haven't tested it though.

Reply:
A bit unfortunate but `cmd!()` only accepts a `literal`, can `concat_str!` output a literal? I don't think so right? And yeah quoting arguments is a problem with the current incorrect behavior of `split_whitespace`.

Reply:
I *could* force the input to be a `&[]`, since

```rs
let args = ["hello", "world"];
let c = cmd!(sh, "echo {args...}");
```

works I think. This however forces the use site to be come something like

```rs
let _output = rustc!(scx, ["--emit=metadata", "--crate-type=lib", "stable.rs"]);
let output = rustc!(
    scx,
    [
        "--emit=metadata",
        "--extern",
        &format!("stable={}/libstable.rmeta", scx.tmpdir()),
        "main.rs"
    ]
);
```

Reply:
Of course. cmd!()` would see the `concat_str!()` before expanding. @matklad would you accept adding something like a method on `Shell` that defines a wrapper program for all commands spawned using this `Shell` instance? Such that this could be `cmd!(sh, $args)` where `sh` is `Shell::new().wrapper_command(rustc_path)`.

Reply:
While a proper fix to this requires a change to `xshell`, I will temporarily use the iterable form for the recipes since they're at least more correct than `split_whitespace`.

Reply:
>let _output = rustc!(scx, ["--emit=metadata", "--crate-type=lib", "stable.rs"]);

We are passing argument as an array here, so I don't think we even need `xshell`? The whole purpose of xshell is to avoid needing an array literal, but, if you are okay with that, I don't think xshell brings much to the table here? 

>would you accept adding something like a method on Shell that defines a wrapper program for all commands spawned using this Shell instance?

This would technically work, but I don't think that's an obvious API with a clear use-case. My answer would be, if you want to control the first arg, to use something like

```
cmd!(sh, "{rustc} --emit=metadata --crate-type=lib")
```

Like shells, xshell supports interpolation of the first argumetn (or at least it should support it).

Reply:
> We are passing argument as an array here, so I don't think we even need `xshell`? The whole purpose of xshell is to avoid needing an array literal, but, if you are okay with that, I don't think xshell brings much to the table here?

(Note that passing-as-an-array here is a temporary band-aid fix, the intention *is* to avoid needing an array literal)